Update Dynamic Instrumentation docs for public beta

OmerRaviv opened the pull request:
### What does this PR do?
Update the Dynamic Instrumentation docs to reflect new features and changes in our UI.

### Motivation
The Dynamic Instrumentation product is transitioning from a private beta to an open beta this week.

<!-- ### Preview -->
<!-- Assuming you are a Datadog employee and named your branch `<yourname>/<description>`, a preview build will run and links to the preview output will be auto-generated and posted in the PR comments. The links will 404 until the preview build is finished running -->


---

### Reviewer checklist
- [ ] Review the changed files.
- [ ] Review the URLs listed in the [Preview](#preview) section.
- [ ] Check images for PII
- [ ] Review any mentions of "Contact Datadog support" for internal support documentation.

shatzi commented on content/en/dynamic_instrumentation/_index.md:
```diff
@@ -114,27 +111,54 @@ For more information, read [How Dynamic Instrumentation Works][11].
+
+1. Select **Span** as the probe type.
+1. Complete the generic probe setup (choose service, environment, version, and probe location). 
+
```
should we create a screen shoot here?

ojung replied:
+1 on the screenshots. Either add them for the new paragraphs or remove them from log and metric probe and maybe just have one exemplary screenshot.

shatzi commented on content/en/dynamic_instrumentation/_index.md:
```diff
@@ -114,27 +111,54 @@ For more information, read [How Dynamic Instrumentation Works][11].
+
+A *span tag* probe decorates an existing span with a tag value.  You can use a *span tag probe* as a more efficient alternative to [using Custom Instrumentation to addings tags in code][14], as it does not require making code changes and redeploying your software.
+
+To create a span probe:
```
To create a span tag probe

OmerRaviv replied:
Fixed in 0e2f1be7a5d7234bc8981a32d5bd14c76c9b18f4

shatzi commented on content/en/dynamic_instrumentation/_index.md:
```diff
@@ -114,27 +111,54 @@ For more information, read [How Dynamic Instrumentation Works][11].
+1. Specify the value of the tag using the [Dynamic Instrumentation expression language][15]. 
+1. Optionally define a condition using the Dynamic Instrumentation expression language. The tag will only be added when the expression evaluates to true.
+1. Optionally add additional tags, each with their own name, expression, and optional condition.
+ 
```
screen shot?

OmerRaviv replied:
added in 0bd29f99cd0db39f0738ffa5261d48e4e883e2b3

shatzi reviewed (approved):
Looks good

ojung commented:
I see some inconsistencies in how we describe log and metrics probes vs span and span tag probes, for example there is still this sentence:

> Both log and metric probes require the same initial setup:

Even though span and span tag probes are the same. This is also showing in the span and span tag probe description which is worded differently than the other probes, while it is essentially the same. Add o11y without making code changes.

ojung commented on content/en/dynamic_instrumentation/_index.md:
```diff
@@ -114,27 +111,54 @@ For more information, read [How Dynamic Instrumentation Works][11].
 
+### Creating a span probe
+
+A *span probe* emits a span when a method is executed. You can use a *span probe* as a more efficient alternative to [creating new spans with Custom Instrumentation]][13], as it does not require making code changes and redeploying your software. If the method throws an exception, the details of the exception will be associated with the newly created span's `error` tag.
```
This doesn't render correctly: 
<img width="836" alt="Screenshot 2023-06-26 at 09 28 32" src="https://github.com/DataDog/documentation/assets/1101441/1de97c59-cace-4fdb-ae85-a2b5325ecf52">

OmerRaviv replied:
Fixed in f76d3d1908e29d59a9366e309684d8b810db3e6c

ojung commented on content/en/dynamic_instrumentation/_index.md:
```diff
@@ -37,17 +32,17 @@ Dynamic Instrumentation requires the following:
 - [Datadog Agent][1] 7.41.1 or higher is installed alongside your service.
 - [Remote Configuration][2] is enabled in that Agent.
-- For Java applications, tracing library [`dd-trace-java`][3] 1.8.0 or higher.
+- For Java applications, tracing library [`dd-trace-java`][3] 1.15.0 or higher.
```
Let's also update minimum `Datadog Agent` version to `7.44.0` as discussed in slack.

OmerRaviv replied:
Done in 6a6e9133b97fffccfd5a4cebcb61b152394dd5e3

ojung commented on content/en/dynamic_instrumentation/how-it-works.md:
```diff
@@ -44,17 +40,69 @@ Dynamic Instrumentation metric probes support the following metric types:
 
+### Span probes
+
+A *span probe* emits a span when a method is executed. You can use a *span probe* as a more efficient alternative to [creating new spans with Custom Instrumentation][4], as it does not require making code changes and redeploying your software. If the method throws an exception, the details of the exception will be associated with the newly created span's `error` tag.
```
This is the exact same as on the other page. For log and metric probes we provide additional information compared to the landing page. Is there anything extra we should add about span (tag) probes?

OmerRaviv replied:
I'm not sure - now that we have more probe types, I'm wondering if having this duplication still makes sense, and if we should consider dropping the "How It Works" page entirely?

ojung replied:
Then we should move any extra information given here to the landing page. If we don't have any extra info for the span probes we could just drop these sections as well. We also definitely need to keep the expression language docs.

OmerRaviv replied:
I've done this refactoring and moved the expression language docs to a different page. WDYT?

OmerRaviv commented:
@ojung 
> This is also showing in the span and span tag probe description which is worded differently than the other probes, while it is essentially the same. Add o11y without making code changes. 

I attempted to address this in f76d3d1908e29d59a9366e309684d8b810db3e6c, couldyou please have a look and let me know if you still see inconsistencies?

ojung commented:
From the how it works page:
> Because capturing this data is performance-intensive, by default it is enabled on only one instance of your service that matches the specified environment and version. Probes with this setting enabled are rate limited to one hit per second.

Should we mention that this behavior can be changed with agent filters? We don't have any section on agent filters anyway, but I think it's worth it to advertise it.

ojung commented:
I think this part

> You can use a span probe as a more efficient alternative to [creating new spans with Custom Instrumentation](https://docs-staging.datadoghq.com/omer.raviv/dynamic-instrumentation-open-beta/tracing/trace_collection/custom_instrumentation/java/#adding-spans), as it does not require making code changes and redeploying your software.

applies to all probe types. Maybe move it up to the general section and drop it from the span probes section and add more  details on the individual probe types, like that a tag can be added on the active or service entry span and shows up in the flame chart and can be used for searching and that a span probe will increase the resolution of your trace and is also visible across the product.

OmerRaviv commented:
@ojung  I addressed your feedback in eedb598e980f35eed2d78920a8d5385bb9b27736, please re-review and lemme know what you think

ojung commented on content/en/dynamic_instrumentation/_index.md:
```diff
@@ -23,118 +23,162 @@ further_reading:
+To enable Dynamic Instrumentation on a service, go to the [in-app setup page][16]. 
 
-To enable Dynamic Instrumentation on a service, select its runtime and follow the setup instructions:
+For more details instructions, select the relevant runtime below:
```
```suggestion
For more detailed instructions, select the relevant runtime below:
```

ojung commented on content/en/dynamic_instrumentation/_index.md:
```diff
@@ -23,118 +23,162 @@ further_reading:
+
+A probe allows you to collect data from specific points in your code without halting the execution of the program. 
+
+If you're a SRE, you can think of probes as a way of enhancing your observability by adding dynamic logs, metrics, and spans to running application without needing to stop or restart anything. This meant you can gather data immediately without disturbing the user experience or requiring lengthy deployments.
```
```suggestion
If you're a SRE, you can think of probes as a way of enhancing your observability by adding dynamic logs, metrics, and spans to running application without needing to stop or restart anything. This means you can gather data immediately without disturbing the user experience or requiring lengthy deployments.
```

ojung commented on content/en/dynamic_instrumentation/_index.md:
```diff
@@ -23,118 +23,162 @@ further_reading:
+If you're a SRE, you can think of probes as a way of enhancing your observability by adding dynamic logs, metrics, and spans to running application without needing to stop or restart anything. This meant you can gather data immediately without disturbing the user experience or requiring lengthy deployments.
 
-Both log and metric probes require the same initial setup:
+If you're a developer, you can think of a probe as a "non-breaking breakpoint". In traditional debugging, a breakpoint is a point in the program where the execution stops, allowing the developer to inspect the state of the program at that point. However, in real-world production environments, it's not practical or even possible to stop the execution of the program. Probes fill in this gap by allo ...
```
Hmm, I'm not sure we should call out the roles here directly. 🤔 I think this is how we think about the personae internally, but do we want to bleed that into the docs? If you're a DevOps, what is DI to you?

OmerRaviv replied:
That's fair, I removed SRE but kept developer in ad899feacc7cef6e2051a3430fc4d5da7daf872b to make this more fluid, does it make sense now?

ojung commented on content/en/dynamic_instrumentation/_index.md:
```diff
@@ -23,118 +23,162 @@ further_reading:
-You can see the captured values in the Datadog UI. 
+If you enable `Capture method parameters and local variables` on the log probe, the following debugging data will be captured and added it to the log event:
+  - **Method arguments**, **local variables**, and **fields**, with the following limits by default:
+    - Follow references three levels deep (configurable in the UI).
```
configurable up to five levels?

OmerRaviv replied:
I think this UX is still up for debate, so prefer leaving that unspecified for now

ojung commented on content/en/dynamic_instrumentation/_index.md:
```diff
@@ -23,118 +23,162 @@ further_reading:
+  - Caught and uncaught **exceptions**.
 
-Because capturing this data is performance-intensive, it is enabled on only one instance of your service that matches the probe's environment and version settings. Probes with capture enabled are rate limited to execute once per second.
+Because capturing this data is performance-intensive, by default it is enabled on only one instance of your service that matches the specified environment and version. Probes with this setting enabled are rate limited to one hit per second.
```
```suggestion
Because capturing the execution context is performance-intensive, by default it is enabled on only one instance of your service that matches the specified environment and version. Probes with this setting enabled are rate limited to one hit per second.
```

ojung commented:
I like it better after the refactor, thanks! Left a couple more suggestions and questions.

OmerRaviv commented on content/en/dynamic_instrumentation/_index.md:
```diff
@@ -23,118 +23,162 @@ further_reading:
+
+A probe allows you to collect data from specific points in your code without halting the execution of the program. 
+
+If you're a SRE, you can think of probes as a way of enhancing your observability by adding dynamic logs, metrics, and spans to running application without needing to stop or restart anything. This meant you can gather data immediately without disturbing the user experience or requiring lengthy deployments.
```
Fellow non-native English speaker, so I solved the problem by working around it in ad899feacc7cef6e2051a3430fc4d5da7daf872b :)

ojung reviewed (approved):
This looks good from my side now, thanks!

EWachnowezki commented on content/en/dynamic_instrumentation/_index.md:
```diff
@@ -23,110 +23,166 @@ further_reading:
+1. Specify a name for the tag.
+1. Specify the value of the tag using the [Dynamic Instrumentation expression language][15]. 
+1. Optionally define a condition using the Dynamic Instrumentation expression language. The tag will only be added when the expression evaluates to true.
+1. Optionally add additional tags, each with their own name, expression, and optional condition.
```
Do we want to say something if the conditions are overlapping: will both tags applied for example or do we stop after the first condition is met?

OmerRaviv replied:
The conditions are per-tag, and I think @shatzi made changes to the UX to make this very clear, so I don't think it necessarily needs explaining - any feedback on this @shatzi ?

shatzi replied:
let look at usage and recording to see if this is a problem. until now I don't see it was an issue and Datadog support engineers help the user to setup on traiil.

EWachnowezki commented on content/en/dynamic_instrumentation/expression-language.md:
```diff
@@ -0,0 +1,63 @@
+* You CAN use comparison operators (`<`, `>`, `>=`, `<=`, `==`, `!=`) to compare variables, fields, and constants in your conditions, as shown in the example `localVar1.field1.field2 != 15.`
+* You CAN use logical operators (`&&`, `||`, and `not` or `!`) to build complex Boolean conditions.
+* You CAN use the `null` literal (equivalent of `nil` in Python)
+* You CANNOT call methods, as Dynamic Instrumentation does not allow users to execute code that may have side effects.
```
```suggestion
* You CANNOT call methods, as Dynamic Instrumentation does not allow users to execute code that may have side effects.
   * Alternatively: You CAN access `private` fields directly

```

OmerRaviv replied:
This suggestion was applied as part of 882852b920ceb87b41906dfaa54a92dce0c61743

kayayarai commented:
tracking docs review in [DOCS-5765](https://datadoghq.atlassian.net/browse/DOCS-5765)

kayayarai reviewed (changes requested):
Looking good! I added in the left-navigation structure and gave it a thorough edit for grammar and style. Please check that I didn't change anything for the worse. Also, I think on line 93-94, we're missing instructions on how to create a probe from within Error Tracking? Plus add any screen caps you want. 

If these changes get in, we can publish this on Thursday.

EWachnowezki commented on content/en/dynamic_instrumentation/expression-language.md:
```diff
@@ -0,0 +1,66 @@
+| `@return`   | Provides access to the return value                                        |
+| `@duration` | Provides access to the call execution duration                             |
+| `@it`       | Provides access to the current value in collection iterating operations    |
+
```
What about @exceptions?

EWachnowezki replied:
We agreed to rather remove it from the UI

EWachnowezki commented on content/en/dynamic_instrumentation/_index.md:
```diff
@@ -23,105 +24,157 @@ further_reading:
+: On a profiler flame graph, you can create a probe for a method by selecting **Instrument this frame with a probe** from the frame's context menu.
+
+Error Tracking
+: 
```
```suggestion
: When you hover inside the stack trace over a stack frame you can click on **Instrument** this pre-fill the probe creation form with the Issue context.
```

OmerRaviv commented:
@kayayarai Thanks so much for your thorough and awesome edits and improvements, we really appreciate it a lot!
We addressed your feedback, and would really love to have this published on Thursday as you mentioned. Is there anything further that's blocking from your perspective?

kayayarai commented on content/en/dynamic_instrumentation/_index.md:
```diff
@@ -91,7 +91,7 @@ Profiling
 
 Error Tracking
-: 
+: When you hover inside the stack trace over a stack frame you can click on **Instrument** this pre-fill the probe creation form with the Issue context.
```
```suggestion
: On a stack trace, mouse over a stack frame and click **Instrument**. This prefills the probe creation form with the Issue context.
```

kayayarai commented:
@OmerRaviv Evgeni asked me to change where it falls in the left-nav, so I'm going to do that and let it rebuild. Otherwise, this is ready to go for tomorrow.
